soc: arm: nordic: handle nRF52832 anomaly 109

JordanYates opened the pull request:
Enable workaround for anomaly 109 by default when affected peripherals are enabled.

JordanYates commented:
Waiting for the underlying problem to be resolved might be the preferred approach: https://devzone.nordicsemi.com/f/nordic-q-a/100913/nrf52832-spis-errata-109-workaround-does-not-work-with-pinctrl

hubertmis commented on soc/arm/nordic_nrf/nrf52/Kconfig.soc:
```diff
@@ -121,4 +121,15 @@ config NRF52_ANOMALY_198_WORKAROUND
+	  Due to Anomaly 109 the first byte sent out by these peripherals is
+	  sometimes wrong. This occurs when the system enters IDLE and stops the
+	  64MHz clock at the same time as the peripheral that is using DMA is started.
+	  This anomaly applies to IC revisions up to "3", the most recent one.
```
"the most recent one" may get outdated and I don't expect anyone to correct this text then.
```suggestion
	  This anomaly applies to IC revisions up to "3".
```

JordanYates replied:
The concern is valid but the same text is present on the anomaly above

anangl commented on dts/bindings/spi/nordic,nrf-spis.yaml:
```diff
@@ -14,3 +14,11 @@ properties:
         Default character. Character clocked out when the slave was not
         provided with buffers and is ignoring the transaction.
+
+  csn-pin-anomaly-109-workaround:
+    type: int
+    description: |
+      Enabling the nRF52832 errata 109 workaround requires the CSN pin to be
+      known by the driver, even when PINCTRL is enabled. There is no way to
+      extract the SPIS_CSN knowledge from the pinctrl node at compile time, so
+      we are forced to specify the value a second time here.
```
This needs be fixed in the nrfx_spis driver. Then this property will not be needed.
Let me provide a PR to hal_nordic that will address this.

anangl replied:
@JordanYates Here's the promised PR: https://github.com/zephyrproject-rtos/hal_nordic/pull/135.

carlescufi replied:
@JordanYates can you please point `west.yml` in this PR to the hal_nordic PR above?

JordanYates replied:
done

awojasinski-nordicsemi replied:
@JordanYates we added this change to the recent nrfx release. Please change revision in manifest to https://github.com/zephyrproject-rtos/hal_nordic/pull/137

anangl commented on modules/hal_nordic/nrfx/nrfx_config.h:
```diff
@@ -164,6 +164,9 @@
 
 #ifdef CONFIG_NRFX_PWM
 #define NRFX_PWM_ENABLED 1
+#ifdef CONFIG_NRF52_ANOMALY_109_WORKAROUND
+#define NRFX_PWM_NRF52_ANOMALY_109_WORKAROUND_ENABLED 1
+#endif
 #endif
```
This double `#endif` doesn't look nice and there is no need to nest those definitions, so I'd suggest to rather use:
```suggestion
#endif
#ifdef CONFIG_NRF52_ANOMALY_109_WORKAROUND
#define NRFX_PWM_NRF52_ANOMALY_109_WORKAROUND_ENABLED 1
#endif
```
or perhaps even better group all those `*_ANOMALY_109_WORKAROUND_ENABLED` symbols in one place:
```
#ifdef CONFIG_NRF52_ANOMALY_109_WORKAROUND
#define NRFX_PWM_NRF52_ANOMALY_109_WORKAROUND_ENABLED 1
#define NRFX_SPIS_NRF52_ANOMALY_109_WORKAROUND_ENABLED 1
#define NRFX_SPIM_NRF52_ANOMALY_109_WORKAROUND_ENABLED 1
#define NRFX_TWIM_NRF52_ANOMALY_109_WORKAROUND_ENABLED 1
#endif
```
This could be updated together with hal_nordic SHA to avoid an extra CI run.

@awojasinski-nordicsemi @nika-nordic What you think?

awojasinski-nordicsemi replied:
I would agree with @anangl. It would be better if you could group all those symbols together

JordanYates replied:
Done

Commit:
soc: arm: nordic: handle nRF52832 anomaly 109

Enable workaround for anomaly 109 by default when affected peripherals
are enabled.

Signed-off-by: Jordan Yates <jordan.yates@data61.csiro.au>

JordanYates commented:
Ping @awojasinski-nordicsemi @anangl

anangl commented on modules/hal_nordic/nrfx/nrfx_config.h:
```diff
@@ -382,6 +382,13 @@
+#define NRFX_SPIM_NRF52_ANOMALY_109_WORKAROUND_ENABLED 1
+#define NRFX_SPIS_NRF52_ANOMALY_109_WORKAROUND_ENABLED 1
+#define NRFX_TWIM_NRF52_ANOMALY_109_WORKAROUND_ENABLED 1
+#define NRFX_PWM_NRF52_ANOMALY_109_WORKAROUND_ENABLED 1
```
Why not in alphabetical order?

fabiobaltieri commented:
Hey hi, I think this is now trying to include a file missing from the hal:

```
$ west build -p -b nrf52dk_nrf52832 samples/basic/fade_led
...
zephyrproject/modules/hal/nordic/nrfx/drivers/src/nrfx_pwm.c:55:10: fatal error: haly/nrfy_egu.h: No such file or directory
   55 | #include <haly/nrfy_egu.h>
      |          ^~~~~~~~~~~~~~~~~
```

Did this depend on a missing HAL update or something?

nika-nordic commented:
@JordanYates FYI: https://github.com/zephyrproject-rtos/zephyr/pull/62658
